File: dynamic_listings/lua/default/computer_science/tree.lua
--[[
-- This is an example of a simple node structure for a tree.
-- It can be used as root or any other node
--]]
Node = {}

function Node:new(o)
    o = o or {
        value = 0,
        left = nil,
        right = nil
    }
    setmetatable(o, self)
    self.__index = self
    return o
end


local function build_example_tree()
    -- Let's build the example tree starting with the nodes
    local A = Node:new({value = "A"})
    local B = Node:new({value = "B"})
    local C = Node:new({value = "C"})
    local D = Node:new({value = "D"})
    local E = Node:new({value = "E"})
    local F = Node:new({value = "F"})
    local G = Node:new({value = "G"})
    local H = Node:new({value = "H"})
    local I = Node:new({value = "I"})
    -- Now we connect the various components (the edges)
    B.left = A
    B.right = C
    F.left = E
    D.left = B
    D.right = F
    H.right = I
    G.left = D
    G.right = H
    -- The tree is ready to be used, let's return the root (G)
    return G
end

File: dynamic_listings/rust/default/computer_science/tree.rs
# #![allow(dead_code)]
/*
 * This is an example of a simple node structure for a tree.
 * It can be used as root or any other node
 */
struct Node {
    content: &'static str,
    left: Option<Box<Node>>,
    right: Option<Box<Node>>,
}

impl Node {
    fn new(content: &'static str) -> Self {
        Self {
            content,
            left: None,
            right: None,
        }
    }
}

fn build_example_tree() -> Node {
    // Let's build the example tree; starting with the nodes
    let a = Node::new("A");
    let mut b = Node::new("B");
    let c = Node::new("C");
    let mut d = Node::new("D");
    let e = Node::new("E");
    let mut f = Node::new("F");
    let mut g = Node::new("G");
    let mut h = Node::new("H");
    let i = Node::new("I");
    // Now we connect the various components (the edges)
    b.left = Some(Box::new(a));
    b.right = Some(Box::new(c));
    f.left = Some(Box::new(e));
    d.left = Some(Box::new(b));
    d.right = Some(Box::new(f));
    h.right = Some(Box::new(i));
    g.left = Some(Box::new(d));
    g.right = Some(Box::new(h));
    // The tree is ready to be used, let's return the root (g)
    g
}
